coprocess/python/dispatcher.py

```diff
@@ -32,58 +31,62 @@ def get_modules(self, the_path):
-                self.middlewares.append(middleware)
-        self.update_hook_table()
+    def find_bundle(self, bundle_id):
+        found = None
+        for bundle in self.bundles:
+            if bundle.bundle_id == bundle_id:
+                found = bundle
+                break
+        return found
+
+    def load_bundle(self, bundle_path):
+        path_splits = bundle_path.split('/')
+        bundle_id = path_splits[-1]
+        bundle = self.find_bundle(bundle_id)
+        if not bundle:
+            bundle = TykMiddleware(bundle_id)
+            self.bundles.append(bundle)
+            self.update_hook_table(with_bundle=bundle)
+
+            return
+        self.update_hook_table(with_bundle=bundle)
 
 
     def load_middlewares(self):
         tyk.log( "Loading middlewares.", "debug" )
-        available_modules = self.get_modules(self.middleware_path)
-        for module_name in available_modules:
-            middleware = self.find_middleware(module_name)
-            if middleware:
-                middleware.reload()
-            else:
-                middleware = TykMiddleware(module_name)
-                self.middlewares.append(middleware)
-        self.update_hook_table()
 
     def purge_middlewares(self):
-        tyk.log( "Purging middlewares.", "debug" )
-        available_modules = self.get_modules(self.middleware_path)
-        for middleware in self.middlewares:
-            if not middleware.filepath in available_modules:
-                tyk.log( "Purging middleware: '{0}'".format(middleware.filepath), "warning" )
-                self.middlewares.remove(middleware)
+        self.middlewares = []
 
-    def update_hook_table(self):
+    def init_hook_table(self):
         new_hook_table = {}
-        for middleware in self.middlewares:
-            for hook_type in middleware.handlers:
-                for handler in middleware.handlers[hook_type]:
-                    handler.middleware = middleware
-                    new_hook_table[handler.name] = handler
+        for bundle in self.bundles:
+            api_id = bundle.api_id
+            hooks = {}
+            for hook_type in bundle.handlers:
+                for handler in bundle.handlers[hook_type]:
+                    handler.middleware = bundle
+                    hooks[handler.name] = handler
+            new_hook_table[api_id] = hooks
         self.hook_table = new_hook_table
 
+    def update_hook_table(self, with_bundle=None):
+        new_hook_table = {}
+        # Disable any previous bundle associated with an API:
+        if with_bundle:
+            # First check if this API exists in the hook table:
+            the_hooks = []
+            if with_bundle.api_id in self.hook_table:
+                the_hooks = self.hook_table[with_bundle.api_id]
+            if len(the_hooks) > 0:
+                # Pick the first hook and get the current bundle:
+                bundle_in_use = list(the_hooks.values())[0].middleware
```

Comment:
Hmm, this doesn't look right — `the_hooks` is a list according to line 78. Unless it becomes a dict on 80, in which case line 78 should be changed to `the_hooks = {}` to avoid confusion.